http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala

```diff
@@ -625,44 +628,66 @@ class DataprocInterpreter[F[_]: Parallel](
-                s"${projectNumber}@cloudservices.gserviceaccount.com"
-              )
-              _ <- updateGroupMembership(config.groupsConfig.dataprocImageProjectGroupEmail,
-                                         apiServiceAccountEmail,
-                                         createCluster
-              )
-            } yield ()
-          }
-      } yield ()
-    }
+    for {
+      count <- inTransaction(clusterQuery.countActiveByProject(CloudContext.Gcp(googleProject)))
+      // Note: Don't remove the account if there are existing active clusters in the same project,
+      // because it could potentially break other clusters. We only check this for the 'remove' case.
+      _ <-
+        if (count > 0 && !createCluster) {
+          F.unit
+        } else {
+          for {
+            projectNumberOpt <- googleResourceService.getProjectNumber(googleProject)
+
+            projectNumber <- F.fromEither(projectNumberOpt.toRight(GoogleProjectNotFoundException(googleProject)))
+            // Note that the Dataproc service account is used to retrieve the image, and not the user's
+            // pet service account. There is one Dataproc service account per Google project. For more details:
+            // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
+
+            // Note we add both service-[project-number]@dataproc-accounts.iam.gserviceaccount.com and
+            // [project-number]@cloudservices.gserviceaccount.com to the group because both seem to be
+            // used in different circumstances (the latter seems to be used for adding preemptibles, for example).
+            // See https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/service-accounts#dataproc_service_accounts_2
+            // for more information
+            dataprocServiceAccountEmail = WorkbenchEmail(
+              s"service-${projectNumber}@dataproc-accounts.iam.gserviceaccount.com"
+            )
+            _ <- updateGroupMembership(config.groupsConfig.dataprocImageProjectGroupEmail,
+                                       dataprocServiceAccountEmail,
+                                       createCluster
+            )
+            _ <-
+              if (createCluster)
+                waitUntilMemberAdded(dataprocServiceAccountEmail)
+              else F.unit
+            apiServiceAccountEmail = WorkbenchEmail(
+              s"${projectNumber}@cloudservices.gserviceaccount.com"
+            )
+            _ <- updateGroupMembership(config.groupsConfig.dataprocImageProjectGroupEmail,
+                                       apiServiceAccountEmail,
+                                       createCluster
+            )
+            _ <-
+              if (createCluster)
+                waitUntilMemberAdded(apiServiceAccountEmail)
+              else F.unit
+          } yield ()
+        }
+    } yield ()
+
+  private def waitUntilMemberAdded(memberEmail: WorkbenchEmail): F[Boolean] = {
```

Comment:
Could you add a docstring to thins function explaining what it does and why? For instance what do the 60 and 5 seconds stand for, and was there any rational behind using these specific values? 
Should they live as constants elsewhere for instance?

Reply:
is the function name clear what it's supposed to do?

> what do the 60 and 5 seconds stand for,
that's documented in `streamUntilDoneOrTimeout`

> was there any rational behind using these specific values
They're my estimates for now. Will need to monitor what happens in tests and ppl's usage to see if adjustment is needed

>Should they live as constants elsewhere for instance?
I don't think that adds too much value